### PR TITLE
MFTF automation for: User saves licensed image into Magento Media Gallery

### PR DESCRIPTION
**Pull request by drpayyne**
### Description

> **~On hold until https://github.com/magento/adobe-stock-integration/issues/555 is done. This is a TDD approach as suggested by @sivaschenko.~**

-  `AdminAdobeStockPreviewedLicensedImageSaveTest`: tests saving a previously licensed image **which already has a preview saved**.
-  `AdminAdobeStockNotPreviewedLicensedImageSaveTest`: tests saving a previously licensed image **which doesn't already have a preview saved**
-  Added `name` attribute for Locate button to obtain previously saved preview image's name in KO template
-  Two tests, since generated image name can be obtained from either prompt field (preview not saved) or Locate button's `name` attribute (preview saved already, so no prompt field while saving licensed image)
-  Both tests delete the licensed image after to return to the state before the test began
-  Fixed a couple of `stepKey` typos

### Fixed Issues
-  magento/adobe-stock-integration#443: [MFTF] User saves licensed image into Magento Media Gallery
-  magento/adobe-stock-integration#442: [MFTF] User views licensed images in the grid

### Manual testing scenarios

-  Add `magento/ADOBE_STOCK_LICENSED_IMAGE` credential with the image ID of a licensed image
-  `$ vendor/bin/mftf run:test AdminAdobeStockNotPreviewedLicensedImageSaveTest --remove`
-  `$ vendor/bin/mftf run:test AdminAdobeStockPreviewedLicensedImageSaveTest --remove`

### CODE REVIEW

File: AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockPreviewNotSavedLicensedImageSaveTest.xml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            <actionGroup ref="AdminOpenMediaGalleryForPageNoEditorActionGroup" stepKey="openMediaGalleryForPage"/>
+            <actionGroup ref="AdminAdobeStockOpenPanelFromMediaGalleryActionGroup" stepKey="openAdobeStockPanel"/>
+            <actionGroup ref="AdminAdobeStockClickSignInActionGroup" stepKey="clickOnSignIn"/>
+            <actionGroup ref="AdminAdobeStockImsPopupSignInFillUserDataActionGroup" stepKey="fillUserCredentials"/>
+            <actionGroup ref="AdminAdobeStockImsPopupClickSignInActionGroup" stepKey="clickSignInImsPopup"/>
+            <actionGroup ref="AdminAdobeStockAssertUserLoggedActionGroup" stepKey="assertUserLoggedIn"/>
+        </before>
+        <after>
+            <actionGroup ref="AdminMediaGalleryDeleteImage" stepKey="deleteImageActionGroup">
+                <argument name="generatedImageName" value="//NEED TO PASS VALUE HERE"/>
```
Comment (drpayyne) [inline AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockPreviewNotSavedLicensedImageSaveTest.xml]: Hi @sivaschenko, I need to pass the generated image name here to delete. And this test is when the preview image is not previously saved. So the generated image name is in another action group (AdminAssertAdobeStockImageSavedActionGroup). I don't think I can accommodate the image delete in that ActionGroup since it is used in multiple places. So do I create a new ActionGroup which will have the same functionality as the one mentioned above, but deletes the image after asserting that it is present in the Media Gallery

File: AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockSection.xml [hunk 2]
```diff
@@ -21,7 +21,8 @@
         <element name="modalError" type="text" selector="//aside[contains(@class, 'modal-popup')]//div[@class='modal-content']//div[text()='Something went wrong.']"/>
         <element name="countGridImages" type="text" selector="//div[@class='masonry-image-grid']/div[contains(.,data-repeat-index)]"/>
         <element name="recordsFound" type="text" selector="//*[@id='adobe-stock-images-search-modal']//*[text()='records found']/parent::div"/>
-        <element name="selectedImagePreview" type="text" selector="//img[contains(@alt,'{{imageName}}')]" parameterized="true"/>
+        <element name="mediaGalleryImage" type="text" selector="//img[contains(@alt,'{{imageName}}')]" parameterized="true"/>
+        <element name="mediaGalleryDeleteButton" type="button" selector="#delete_files"/>
```
Comment (drpayyne) [inline AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockSection.xml:25]: @sivaschenko, I think these _can_ be moved to another Section since these two are not really related to Adobe Stock, but rather the Magento Media Gallery.

File: AdobeStockImageAdminUi/view/adminhtml/web/js/components/grid/column/preview/actions.js [hunk 3]
```diff
@@ -351,6 +351,15 @@ define([
...
             return this.isDownloaded() ?  $.mage.__('License') : $.mage.__('License and Save');
+        },
+
+        /**
+         * Returns image name of previously saved preview image
+         *
+         * @returns {String}
+         */
+        getImageName: function () {
+            return this.preview().displayedRecord()['path'].split('/').pop();
```
Comment (drpayyne) [inline AdobeStockImageAdminUi/view/adminhtml/web/js/components/grid/column/preview/actions.js]: @sivaschenko, this logic is based on an assumption and precaution that images can be nested in folders, so will have multiple directory slashes in the path. I can modify it to `return this.preview().displayedRecord()['path'].substring(1);` to remove the leading slash if necessary.

**Comment by drpayyne** [comment]
For `AdminAdobeStockPreviewSavedLicensedImageSaveTest`, as discussed with @sivaschenko: https://github.com/magento/adobe-stock-integration/issues/446#issuecomment-549685367, _Locate_ button should be visible when image preview is saved and image is licensed in Adobe Stock (or anywhere else).

But this PR https://github.com/magento/adobe-stock-integration/pull/646/files#diff-d2bf191f817963fae506090c774bdd35R21 changes the visibility of the _Locate_ button to be visible only if image is downloaded AND NOT licensed.

This test is for saved preview image which is licensed elsewhere and according to @sivaschenko and myself, _Locate_ button should be present here. 

So, I'm needing some guidance here. Thanks!

File: AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockSection.xml (hunk 2 above)
Comment (sivaschenko) [reply AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockSection.xml:25]: @drpayyne that's a really good suggestion! Ideally, these selectors can be contributed to the CMS module of magento2 repository, but currently we have to keep them in our module. Moving to a separate section would be good.

File: AdobeStockImageAdminUi/view/adminhtml/web/js/components/grid/column/preview/actions.js (hunk 3 above)
Comment (sivaschenko) [reply AdobeStockImageAdminUi/view/adminhtml/web/js/components/grid/column/preview/actions.js]: @drpayyne Images really can be nested in folders. Image are saved to Media Gallery folder from which Adobe Stock Integration panel is opened.

**Comment by sivaschenko** [comment]
"But this PR https://github.com/magento/adobe-stock-integration/pull/646/files#diff-d2bf191f817963fae506090c774bdd35R21 changes the visibility of the Locate button to be visible only if the image is downloaded AND NOT licensed." - @drpayyne  good catch, thank you, this issue is fixed by https://github.com/magento/adobe-stock-integration/pull/649

**Comment by diazwatson** [comment]
Git conflicts are now fixed

**Comment by diazwatson** [comment]
I am getting the following error when running the test
```
Time: 31.07 seconds, Memory: 6.00MB

There was 1 error:

---------
1) AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest: Admin adobe stock preview not saved licensed image save test
 Test  tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:AdminAdobeStockPreviewNotSavedLicensedImageSaveTest

  [MalformedLocatorException] CSS or XPath locator is malformed: {{AdobeStockSection.mediaGalleryImage({{generatedImageName}})}}


Scenario Steps:

 18. $I->seeElement("{{AdobeStockSection.mediaGalleryImage({{generatedImage...") at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:82
 17. // Entering Action Group [deleteImageActionGroup] AdminMediaGalleryDeleteImage
 16. $I->saveScreenshot() at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:119
 15. $I->waitForElementVisible(".scalable.action-add-image.plugin",10) at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:43
 14. $I->click("div[data-index=content]") at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:42
 13. $I->waitForPageLoad(30) at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:41

#1  Codeception\Module\WebDriver->seeElement
#2  /Users/rdiaz/Sites/adobestock/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/_generated/AcceptanceTesterActions.php:1932
#3  /Users/rdiaz/Sites/adobestock/dev/tests/acceptance/tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest.php:82
#4  Magento\AcceptanceTest\_default\Backend\AdminAdobeStockPreviewNotSavedLicensedImageSaveTestCest->_after

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

**Comment by filmaj** [comment]
One question that needs tackling in here (as well as in #639) is the preconditions around the Adobe Stock account and its (pre)-licensed images. That is, what kind of assumptions can we make about the Adobe Stock account we will be using during testing? Will it come "seeded" with previously-licensed images? If not, will we have the ability to license images (and possibly incur costs on the account) at runtime on every test execution? It seems like this PR (and #639) expect the former situation, and require us hard-coding the licensed image ID ahead of time in an environment variable. This is possibly something we need to bring up with the Adobe Stock team.

**Comment by drpayyne** [comment]
@filmaj True, and it is actually a block in two issues I'm assigned to. #440 and #442. Need some guidance on how to obtain a new image for every test run.

**Comment by drpayyne** [comment]
Hi @diazwatson. Sorry about that. I overlooked that mistake. I've fixed it now. Will test locally and get back soon. Thank you.

File: AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockPreviewNotSavedLicensedImageSaveTest.xml (hunk 1 above)
Comment (drpayyne) [reply AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockPreviewNotSavedLicensedImageSaveTest.xml]: @filmaj could I get some guidance here, please? Thank you.

**Comment by drpayyne** [comment]
Hey @sivaschenko and @diazwatson (@filmaj too if you're interested), couple of thoughts on these two tests regarding maintaining the Magento state.
`AdminAdobeStockPreviewNotSavedLicensedImageSaveTest`: After the test, will be deleting the saved image, among other things like logging out. And the state will be maintained since there was no image saved prior to the test (neither preview image nor licensed image).
`AdminAdobeStockPreviewSavedLicensedImageSaveTest`: Same steps as above to run after the test. Although deleting the saved licensed image, would also delete the preview image. Before the test, a preview image exists but after the test, no image exists. And doesn't fulfill the consistent state of Magento since test will fail on the subsequent run and this (functionally) transforms into the previous test mentioned above which has no preview image saved before the test.

Any suggestions on how to tackle this scenario?

Apologies for the long explanation, hope I was clear. If not, do let me know, I could demo it. Also note, I've changed test names and environment variable names for clarity from the previous commits. 

Thanks.

File: AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        <actionGroup ref="AdminSearchImagesOnModalActionGroup" stepKey="searchForNotPreviewedLicensedImageToSave">
+            <argument name="query" value="{{AdobeStockConfigDataNotPreviewedLicensedNotSavedImage.value}}"/>
+        </actionGroup>
+        <actionGroup ref="AdminAdobeStockExpandImagePreviewActionGroup" stepKey="expandLicensedImage"/>
+        <click selector="{{AdobeStockImagePreviewSection.saveLicensedImage}}" stepKey="clickSave"/>
+        <waitForPageLoad stepKey="waitForPromptModal" />
+        <grabValueFrom selector="{{AdobeStockImagePreviewSection.generatedImageName}}" stepKey="grabImageFileName" />
+        <click selector="{{AdobeStockImagePreviewSection.confirm}}" stepKey="clickOnPopupConfirm"/>
+        <waitForPageLoad stepKey="waitForMediaGalleryOpen" />
+        <seeElement selector="{{AdobeStockSection.mediaGalleryImage({$grabImageFileName})}}" stepKey="assertSavedImage" />
```
Comment (drpayyne) [inline AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml]: @diazwatson I used direct code here instead of the action group: `AdminAssertAdobeStockImageSavedActionGroup` as this test and the other test in this PR have the same selected actions, but in a slightly different order. So I'm wondering if I can decouple that action group, and reorder as necessary.

**Comment by filmaj** [comment]
@drpayyne I think all tests should assume that there are no saved images in the media gallery, before and after test runs. So with that in mind, tests should ensure saving/deleting images as needed in the `before`/`after` blocks. What do you think?

**Comment by drpayyne** [comment]
@filmaj sounds perfect. I'll implement that. Thank you!

**Comment by diazwatson** [comment]
Hi @drpayyne how are you doing with the requested changes, is there anything we can help with?

**Comment by drpayyne** [comment]
Hi @diazwatson, thank you. I'm going to proceed with what @filmaj suggested. For `AdminAdobeStockNotPreviewedLicensedImageSaveTest`, I'll go ahead as usual and for `AdminAdobeStockPreviewedLicensedImageSaveTest`, I'll use the same licensed image identifier, but will perform the _Save Preview_ flow in the `<before/>` node, so once the test starts, the preview is saved. And for both these tests, I'll delete the image file in the `<after/>` node. So that the state is maintained and consistent across multiple test runs.

**Comment by drpayyne** [comment]
I've updated the code according to my comment above, @diazwatson. A thought: https://github.com/magento/adobe-stock-integration/pull/637#discussion_r343476418

File: AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml (hunk 4 above)
Comment (drpayyne) [reply AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml]: This is still an issue regarding the DRY principle. I can refactor `AdminAssertAdobeStockImageSavedActionGroup` into two action groups, and reuse them here, IF it is possible to pass `stepKey` data from one action group to another (need to pass image file name from the first action group to the other). So the above code would become:
```
<!-- The decoupled action groups -->
<actionGroup ref="AdminAdobeStockGrabImageNameActionGroup" name="grabImageName"/>
<actionGroup ref="AdminAssertAdobeStockImageSavedActionGroup" name="assertImageSave"/>
```
instead of the original coupled flow
```
<!-- The coupled action groups -->
<actionGroup ref="AdminAssertAdobeStockImageSavedActionGroup" name="assertImageSave"/>
```

File: AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockSection.xml (hunk 2 above)
Comment (filmaj) [reply AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockSection.xml:25]: @sivaschenko @drpayyne should I create an issue for moving the selectors out of this project and into magento2? Possibly as part of Milestone 3 (after the next 2.3.4 release)?

**Comment by filmaj** [comment]
I'm trying this test w/ the latest develop branch and I'm not seeing images that I _think_ should be recognized as licensed showing up as licensed in the Stock search grid. Data returned from the Stock API in my case has empty data for the  `is_licensed` property.

Screen Shot 2019-11-07 at 1 07 41 PM

So tough for me to test this at this moment 😞

**Comment by drpayyne** [comment]
@filmaj Can you check if it is licensed in the License History in stock.adobe.com?

File: AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml (hunk 4 above)
Comment (rogyar) [reply AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml]: I would suggest making the mentioned actions. Instead of using the direct actions in the test, it's better to create two action groups:
- `AdminAdobeStockGrabImageNameForSaveActionGroup`
- `AdminAssertImageInMediaGalleryActionGroup`

Please note the name of the second action group. It's not something related to AdobeStock, but the standard media gallery. There's a chance that something similar already exists, I would recommend to double-check it.

Comment (drpayyne) [reply AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockNotPreviewedLicensedImageSaveTest.xml]: Thanks, @rogyar! I shall implement these changes.

**Comment by rogyar** [comment]
Hi @drpayyne. Just a suggestion. Please, try to use only action groups in your tests and avoid direct operations call. It's one of the key points of MFTF Coverage Best practices. In that way, we create reusable components that might be helpful for other tests as well

**Comment by drpayyne** [comment]
@rogyar yes, of course. I merely wanted to ask before decoupling an existing action group into two action groups for this particular scenario. Thanks!

**Comment by filmaj** [comment]
I'm having the exact same issue as in #639 when running the test:

```
[searchForNotPreviewedLicensedImageToSave] AdminSearchImagesOnModalActionGroup
  [fillQueryInput] fill field "input.data-grid-search-control","193210330"
  [applySearch] click "#words + button.action-submit"
  [waitForSearchForQuery] wait for loading mask to disappear
[expandLicensedImage] AdminAdobeStockExpandImagePreviewActionGroup
  [clickOnThumbnail] click "aside.adobe-stock-modal [data-role=grid-wrapper] img[data-role=thumbnail]"
  [waitForImagePreviewToExpand] wait for loading mask to disappear
[clickSave] click "//button[@class='action-default primary']//span[text()='Save']"
[saveScreenshot] save screenshot
```

File: AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockImagePreviewSection.xml [hunk 5]
```diff
@@ -8,7 +8,9 @@
 
 <sections xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:mftf:Page/etc/SectionObject.xsd">
     <section name="AdobeStockImagePreviewSection">
-        <element name="save" type="block" selector="//button[@class='action-secondary']/span[contains(.,'Save Preview')]"/>
+        <element name="save" type="block" selector="//button[@class='action-secondary']//span[text()='Save Preview']"/>
+        <element name="locateImage" type="block" selector="//button[@class='action-secondary']//span[text()='Locate')]"/>
+        <element name="saveLicensedImage" type="button" selector="//div[@class='actions']/descendant::span[text()='Save']"/>
```
Comment (drpayyne) [inline AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockImagePreviewSection.xml:13]: Hi @filmaj, please refer: https://github.com/magento/adobe-stock-integration/pull/639#discussion_r346739377 as it is the same issue and fix.

**Comment by drpayyne** [comment]
Although there is one problem which I just realized. Since we shouldn't be licensing new images for every test run (https://github.com/magento/adobe-stock-integration/issues/440#issuecomment-550007076), I use a prelicensed image in the credentials.
-  `AdminAdobeStockNotPreviewedLicensedImageSaveTest`: test is successful as the test should ensure a licensed but not previewed image _can_ be saved.
-  `AdminAdobeStockPreviewedLicensedImageSaveTest`: test is unsuccessful and I think not possible as, a licensed image will have just one button: _Save_. It isn't possible to save a preview of an already licensed image in the `<before/>` node. So this flow cannot be tested unless we take a random unlicensed image, save the preview, and license the image in the `<before/>` node, and then test if we can save this image.

**Comment by filmaj** [comment]
We need to make sure the appropriate environment variable is set in travis' settings.

**Comment by filmaj** [comment]
I have added the env var to travis settings, FYI
